## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [28]:
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [49]:
# View the column names in the data
columns = df.columns
print(columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'shipping_price', 'line_cost', 'subtotal',
       'total_price'],
      dtype='object')


In [56]:
# Use the describe function to gather some basic statistics
stats = df.describe(include='all')
print(stats)

              first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      o

In [37]:
df.tail(20)

first       last                                      job  \
54619      Laura      Owens                        Buyer, industrial   
54620       Mary     Wilson               Further education lecturer   
54621      Bryan     Reeves                      Solicitor, Scotland   
54622     Alyssa   Martinez                Education officer, museum   
54623     Gloria    Navarro                   Audiological scientist   
54624  Elizabeth       Cruz                Commercial horticulturist   
54625      Danny    Jackson                             Toxicologist   
54626   Danielle      Moore                        Financial manager   
54627       Nina     Mullen                            Warden/ranger   
54628    Anthony     Tucker      Designer, blown glass/stained glass   
54629      Linda  Henderson  Scientist, research (physical sciences)   
54630       Lisa     Riddle          Chartered management accountant   
54631      Kevin     Obrien                           Microbiologist   
54632     Austin     Fuller                       Investment analyst   
54633     Tamara      Jones                        Fish farm manager   
54634     Lauren      Reese                           Radio producer   
54635    Derrick      Moore                                 Musician   
54636     Monica  Gutierrez                         Graphic designer   
54637      Wanda    Solomon                             Toxicologist   
54638     Dwayne       Bell                  Chief Financial Officer   

                       phone                                    email  \
54619           753-596-3400                 lauraowens6281@baker.org   
54620           307.853.9158                   wilsonm8454@bender.net   
54621    +1-515-711-5348x987        reeves.participant.7298@smith.com   
54622           440-527-7555              a.martinez.218@thompson.com   
54623          (481)651-9381               navarrogloria2626@lane.com   
54624     311-227-5670x39674               elizabeth.cruz@mercado.com   
54625  +1-959-625-3646x01410  danny.jackson.817@sherman-armstrong.net   
54626     (932)238-4205x1418        mooreteam3804@mcclain-aguilar.biz   
54627  001-595-635-1880x1626                  mullennina3963@fox.info   
54628        +1-503-460-2669                       tuckera@nguyen.com   
54629    +1-664-405-4100x443             lhenderson@boyer-krueger.com   
54630       001-932-725-9686            riddlel1336@hamilton-cole.net   
54631             7112448766             line.kevin@hart-williams.com   
54632           379.662.7549               fulleraustin7494@nunez.com   
54633       375.495.3124x931      tamara.jones.9326@george-carter.com   
54634     977.877.5272x11382       late_reese_4081@montoya-chavez.org   
54635           358.661.5483             derrick.moore.2602@pope.info   
54636       294.805.9100x339                gutierrezm3195@morris.org   
54637          (311)767-4924                solomonwanda5962@ross.org   
54638      606-619-7667x9508                         belld@galvan.com   

       client_id  order_id  order_date  order_week  order_year  \
54619      27771   8016714  2023-05-13          19        2023   
54620      69074   3730867  2023-05-19          20        2023   
54621      65687   3521909  2023-05-04          18        2023   
54622      47752   6530543  2023-01-17           3        2023   
54623      89110   9012482  2023-02-05           5        2023   
54624      49457   1696742  2023-02-24           8        2023   
54625      84156   4925927  2023-02-15           7        2023   
54626      83552   8673573  2023-03-23          12        2023   
54627      71958   3711021  2023-04-05          14        2023   
54628      32500   2795442  2023-01-05           1        2023   
54629      79817   6473704  2023-03-16          11        2023   
54630      19904   6871726  2023-03-18          11        2023   
54631      39939   6430193  2023-04-07          14        2023   
54632      50317   5064968  2023-05-26  

In [ ]:
# Use this space to do any additional research
# and familiarize yourself with the data.


#the dataset is tracking clients and their purchase histories across categories.  
# this means we can use this data to understand buying patterns, client segmentation, and sales trends. 
#the dataset contains numbers and also strings making it difficult to just blindly run stats functions 
# across the whole data set without specifying the columns first
# there is a wide variety of occupations, and a very big differential in price/item. the big swings
# in average order value make calculating a true aov (average order volume) challenging without providing
#additional context (i.e. focusing it by price brackets or customer type or by category type)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   first           54639 non-null  object 
 1   last            54639 non-null  object 
 2   job             54639 non-null  object 
 3   phone           54639 non-null  object 
 4   email           54639 non-null  object 
 5   client_id       54639 non-null  int64  
 6   order_id        54639 non-null  int64  
 7   order_date      54639 non-null  object 
 8   order_week      54639 non-null  int64  
 9   order_year      54639 non-null  int64  
 10  item_id         54639 non-null  object 
 11  category        54639 non-null  object 
 12  subcategory     54639 non-null  object 
 13  unit_price      54639 non-null  float64
 14  unit_cost       54639 non-null  float64
 15  unit_weight     54639 non-null  float64
 16  qty             54639 non-null  int64  
 17  line_number     54639 non-null 

In [ ]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print(top_categories)



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [ ]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = df['category'].value_counts().idxmax()
subcategory_counts = df[df['category'] == most_common_category]['subcategory'].value_counts().head(1)
print(f"Most common category: {most_common_category}")
print(subcategory_counts)

# alt: df.loc[df.category == 'consumables', 'subcategory'].value_counts().head(1)


Most common category: consumables
subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [ ]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print(top_clients)

#alt: df.client_id.value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [57]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df.client_id.value_counts().head(5).index.tolist()
top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [ ]:
# How many total units (the qty column) did the client with the most entries order order?
most_entries_client = top_clients.index[0]
total_units = df[df['client_id'] == most_entries_client]['qty'].sum()
print(f"Client {most_entries_client} ordered a total of {total_units} units.")

#alt: df.loc[df.client_id == 33615, 'qty'].sum()

Client 33615 ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [58]:
# Create a new column 'subtotal' by multiplying 'unit_price' and 'qty'
df['subtotal'] = df['unit_price'] * df['qty']
df[['unit_price', 'qty', 'subtotal']].head()


unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [59]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.


def shipping_calculate(total_weight):
    if total_weight > 50:
        return total_weight * 7
    else:
        return total_weight * 10

        
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(shipping_calculate)
df.head()




#question: it seems as if each row of data is isolated by the specific item purchased, and 
# that a customer who placed a multi-item order will be entered in the database as 1 entry per item contained in the order
# if this is correct, then calculating the shipping price should be done at the ORDER level, not the unit level
# and this particular calculation could possibly result in overcalculating shipping costs PER ORDER. Is that true?

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  line_cost   subtotal  total_price    profit  total_weight  
0         5512.50   80159.55  115164.00    125891.67  45732.12        787.50  
1          312.90     331.79     523.95       587.32    255.53         31.29  
2          458.64     323.34     527.28       592.85    269.51         65.52  
3          356.70     732.95    1056.18      1166.18    433.23         35.67  
4         6500.20    2627.70    3902.00      4727.24   2099.54        928.60  

[5 rows x 24 columns]

In [60]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.095
df.head()


df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  line_cost   subtotal   total_price    profit  total_weight  
0         5512.50   80159.55  115164.00  132140.76750  45732.12        787.50  
1          312.90     331.79     523.95     916.35075    255.53         31.29  
2          458.64     323.34     527.28    1079.58240    269.51         65.52  
3          356.70     732.95    1056.18    1547.10360    433.23         35.67  
4         6500.20    2627.70    3902.00   11390.40900   2099.54        928.60  

[5 rows x 24 columns]

In [61]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()





first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  line_cost   subtotal   total_price    profit  total_weight  
0         5512.50   85597.05  115164.00  132140.76750  45732.12        787.50  
1          312.90     629.79     523.95     916.35075    255.53         31.29  
2          458.64     765.18     527.28    1079.58240    269.51         65.52  
3          356.70    1077.35    1056.18    1547.10360    433.23         35.67  
4         6500.20    8663.60    3902.00   11390.40900   2099.54        928.60  

[5 rows x 24 columns]

In [62]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = (df['total_price'] - df['line_cost'])
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  line_cost   subtotal   total_price       profit  \
0         5512.50   85597.05  115164.00  132140.76750  46543.71750   
1          312.90     629.79     523.95     916.35075    286.56075   
2          458.64     765.18     527.28    1079.58240    314.40240   
3          356.70    1077.35    1056.18    1547.10360    469.75360   
4         6500.20    8663.60    3902.00   11390.40900   2726.80900   

   total_weight  
0        787.50  
1         31.29  
2         65.52  
3         35.67  
4        928.60  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above

# List of specific order IDs to calculate the total price for
order_ids = [2742071, 2173913, 6128929]

# Filter the DataFrame for only the selected order IDs and make a copy
selected_orders = df[df['order_id'].isin(order_ids)].copy()

# Sum the 'subtotal' and 'shipping_price' per order_id to get the total for each order
order_totals = selected_orders.groupby('order_id').agg({
    'subtotal': 'sum',
    'shipping_price': 'sum'
}).reset_index()

# Calculate the total price for each of these orders using the formula
order_totals['total_price'] = (order_totals['subtotal'] + order_totals['shipping_price']) * 1.0925

# Display the order IDs and their corresponding total prices
print(order_totals[['order_id', 'total_price']])

# Verify with the receipts
expected_prices = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Compare the calculated total prices with the expected values
for order_id in order_ids:
    calculated = order_totals[order_totals['order_id'] == order_id]['total_price'].values[0]
    expected = expected_prices[order_id]
    print(f"Order ID {order_id}:")
    print(f"Calculated Total Price: ${calculated:.2f}")
    print(f"Expected Total Price: ${expected:.2f}")
    print(f"Difference: ${calculated - expected:.2f}\n")


   order_id    total_price
0   2173913  162388.719300
1   2742071  152811.897075
2   6128929  923441.240450
Order ID 2742071:
Calculated Total Price: $152811.90
Expected Total Price: $152811.89
Difference: $0.01

Order ID 2173913:
Calculated Total Price: $162388.72
Expected Total Price: $162388.71
Difference: $0.01

Order ID 6128929:
Calculated Total Price: $923441.24
Expected Total Price: $923441.25
Difference: $-0.01



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [68]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

# find top 5 clients
top_client_data = df[df['client_id'].isin(top_5_clients)]

# calculate total spend for each of the top 5
top_5_total_spend = top_client_data.groupby('client_id')['total_price'].sum().reset_index()

# print the total spending by each of the top 5 clients
print(top_5_total_spend[['client_id', 'total_price']])


   client_id   total_price
0      24741  8.245715e+07
1      33615  8.396479e+06
2      38378  1.293609e+07
3      46820  9.766091e+06
4      66037  1.028299e+07


In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.

# Step 1: Filter the DataFrame for only the top 5 clients
top_clients_data = df[df['client_id'].isin(top_5_clients)]

# Step 2: Calculate the required totals for each client
summary = top_clients_data.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()


# Step 3: Print the summary DataFrame
print(summary)

# Format the data and rename the columns to names suitable for presentation.
# Define the money columns.
money_columns =['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return f"${x / 1_000_000:.2f}M" if pd.notnull(x) else "$0.00M"

# Apply the currency_format_millions function to only the money columns.
summary[money_columns] = summary[money_columns].applymap(currency_format_millions)

# Rename the columns to reflect the change in the money format.
summary.rename(columns={
               'total_units_purchased': 'Total Units Purchased',
               'total_shipping_price': 'Total Shipping Price (in millions)',
               'total_revenue': 'Total Revenue (in millions)',
               'total_profit': 'Total Profit (in millions)'
               }, inplace=True)
print(summary)

   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862            5126448.37   8.245715e+07   
1      33615                  64313            1828984.89   8.396479e+06   
2      38378                  73667            3429455.40   1.293609e+07   
3      46820                  75768            1601448.84   9.766091e+06   
4      66037                  43018            1395151.85   1.028299e+07   

   total_profit  
0  3.676825e+07  
1  2.221165e+06  
2  3.301364e+06  
3  2.758900e+06  
4  3.278509e+06  
   client_id  Total Units Purchased Total Shipping Price (in millions)  \
0      24741                 239862                             $5.13M   
1      33615                  64313                             $1.83M   
2      38378                  73667                             $3.43M   
3      46820                  75768                             $1.60M   
4      66037                  43018                             $

/var/folders/x6/0lvp4qbd3qj2npx_kwg08vh00000gn/T/ipykernel_39795/3840025298.py:22: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary[money_columns] = summary[money_columns].applymap(currency_format_millions)


In [81]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

sorted_summary = summary.sort_values(by='Total Profit (in millions)', ascending=False)
print(sorted_summary)


   client_id  Total Units Purchased Total Shipping Price (in millions)  \
0      24741                 239862                             $5.13M   
2      38378                  73667                             $3.43M   
4      66037                  43018                             $1.40M   
3      46820                  75768                             $1.60M   
1      33615                  64313                             $1.83M   

  Total Revenue (in millions) Total Profit (in millions)  
0                     $82.46M                    $36.77M  
2                     $12.94M                     $3.30M  
4                     $10.28M                     $3.28M  
3                      $9.77M                     $2.76M  
1                      $8.40M                     $2.22M  


In [ ]:
#for this exercise, I manipulated and analyzed a dataset that captured what products customers had ordered from a company.
# I created some totals to help simplify analysis, created a way to calculate shipping costs for each order and profit for each order, 
# and grouped this data to identify top customers and understand their total value and purchase patterns.
# Understanding this data helps understand just how valuable the top customers are (and comparing this against the bottom
# tier of customers helps understand where the team should focus from a marketing/sales standpoint), and also which products
# are most valuable to the most valuable customers (which can help inform future product development/merchandising).
